packages/core/src/impl-store.ts

```diff
@@ -223,28 +226,37 @@ async function checkForAddressClash(
   provider: EthereumProvider,
   data: ManifestData,
   updated: Deployment & RemoteDeploymentId,
-  checkAllAddresses: boolean,
+  merge: boolean,
 ): Promise<void> {
-  const clash = lookupDeployment(data, updated.address, checkAllAddresses);
-  if (clash !== undefined) {
-    if (await isDevelopmentNetwork(provider)) {
+  let clash;
+  const findClash = () => lookupDeployment(data, updated.address, !merge);
+
+  if (await isDevelopmentNetwork(provider)) {
+    // Look for clashes so that we can delete deployments from older runs.
+    // `merge` only checks primary addresses for clashes, since the address could already exist in an allAddresses field
+    // but the updated and stored objects are different instances representing the same entry.
+    clash = findClash();
+    if (clash !== undefined) {
       debug('deleting a previous deployment at address', updated.address);
       clash.set(undefined);
-    } else {
+    }
+  } else if (!merge) {
+    clash = findClash();
+    if (clash !== undefined) {
       const existing = clash.get();
       // it's a clash if there is no deployment id or if deployment ids don't match
       if (
         existing !== undefined &&
         (existing.remoteDeploymentId === undefined || existing.remoteDeploymentId !== updated.remoteDeploymentId)
       ) {
         throw new Error(
-          `The following deployment clashes with an existing one at ${updated.address}\n\n` +
-            JSON.stringify(updated, null, 2) +
-            `\n\n`,
+          `The deployment clashes with an existing one at ${updated.address}.\n\n` +
+            `Existing deployment: ${JSON.stringify(existing, null, 2)}\n\n` +
+            `New deployment: ${JSON.stringify(updated, null, 2)}\n\n`,
```

Comment:
I am not sure printing both JSON's is ok, it's difficult to read in the CLI, what if?

```suggestion
            `Deployments: ${JSON.stringify({ existing, updated }, null, 2)}\n\n`,
```